### PR TITLE
Add Position metadata to the parser

### PR DESCRIPTION
**Pull request by slrtbtfs**
This aims to implement part one of proposal <https://github.com/slrtbtfs/promql-lsp/blob/master/doc/proposals/2019_promql-parser-improvements.md>

It introduces position metadata in the error messages and the AST.

Current state: 
-  For error messages, the implementation is done.
- ~The AST is able to store position metadata but currently mostly uses zeros as placeholders.~
- ~The last commit fills some of the position fields with real data. Unfortunately it triggers a data race. Despite staring at the code for quite a while I haven't been able to figure out the root cause of that race, so any help there would be appreciated.~

### CODE REVIEW

**Comment by brancz** [comment]
We looked at the code a little and couldn't find the data race, maybe someone sees it or can think of a data race that may had been there all along, but only now triggers the race detector.

**Comment by brancz** [comment]
cc @simonpasquier @krasi-georgiev @brian-brazil @juliusv

**Comment by slrtbtfs** [comment]
Output of the race detector: https://gist.github.com/slrtbtfs/359c05c7a5f28c1df3f9d9a1b6be99c6

**Comment by slrtbtfs** [comment]
Data race has been found by Casey and Simon approx. at the same time.

**Comment by slrtbtfs** [comment]
Update: All AST Nodes have Postion information now. There are still two edge cases where that the positions are not accurate:

- NumberLiterals that are the right hand side of a vector matching report positions that are off by one.
- Offset modifiers are not handled correctly.

File: promql/parse.go [hunk 1]
```diff
@@ -80,15 +77,16 @@ func ParseMetric(input string) (m labels.Labels, err error) {
...
 	p := newParser(input)
+	pos := token.Pos(1)
 	defer p.recover(&err)
 
 	name := ""
 	if t := p.peek().typ; t == ItemMetricIdentifier || t == ItemIdentifier {
 		name = p.next().val
 	}
-	vs := p.VectorSelector(name)
+	vs := p.VectorSelector(name, pos)
```
Comment (simonpasquier) [inline promql/parse.go]: it could be `vs := p.VectorSelector(name, token.Pos(1))`

File: promql/printer.go [hunk 2]
```diff
@@ -137,7 +127,14 @@ func (node *BinaryExpr) String() string {
 }
 
 func (node *Call) String() string {
-	return fmt.Sprintf("%s(%s)", node.Func.Name, node.Args)
+	args := ""
+	for _, e := range node.Args {
+		args += e.String()
+		args += ", "
+	}
+	args = args[:len(args)-2]
```
Comment (simonpasquier) [inline promql/printer.go]: it panics for `time()`

File: promql/engine.go [hunk 3]
```diff
@@ -358,6 +359,10 @@ type testStmt func(context.Context) error
 func (testStmt) String() string { return "test statement" }
 func (testStmt) stmt()          {}
 
+// FIXME
```
[the remark this answers is not in the document]
Comment (squat) [reply promql/engine.go]: Add note about fake nodes in AST

File: promql/ast.go [hunk 4]
```diff
@@ -74,18 +88,24 @@ type Expr interface {
 	expr()
 }
 
-// Expressions is a list of expression nodes that implements Node.
-type Expressions []Expr
```
[the remark this answers is not in the document]
Comment (slrtbtfs) [reply promql/ast.go]: The reason why this type was there, is that Expressions used to implement the Node interface, and thus was more than just a list of Expr.

The Problem with that is, that in the case of an empty list of expressions, e.g. the list of Arguments of `time()` there is no reasonable way to figure out it's position.

Now it doesn't implement any interfaces and the type doesn't have a reason for existence any more, so I removed it.

File: promql/lex.go [hunk 5]
```diff
@@ -388,40 +400,51 @@ func (l *lexer) acceptRun(valid string) {
...
 
-// linePosition reports at which character in the current line
-// we are on.
-func (l *lexer) linePosition() int {
-	lb := strings.LastIndex(l.input[:l.lastPos], "\n")
-	if lb == -1 {
-		return 1 + int(l.lastPos)
-	}
-	return 1 + int(l.lastPos) - lb
+// Pos reports the position we're on, based on the position of
```
[the remark this answers is not in the document]
Comment (slrtbtfs) [reply promql/lex.go]: These are two distinct data types, which can be converted into each other:

    type Pos int

    type Position struct {
        Filename string // filename, if any
        Offset   int    // offset, starting at 0
        Line     int    // line number, starting at 1
        Column   int    // column number, starting at 1 (byte count)
    }

`Pos()` returns a `Pos` so the name actually makes sense.

File: promql/lex.go [hunk 6]
```diff
@@ -388,40 +400,51 @@ func (l *lexer) acceptRun(valid string) {
...
 func lex(input string) *lexer {
+	fileSet := token.NewFileSet()
+	file := fileSet.AddFile("", -1, len(input))
+	file.SetLinesForContent([]byte(input))
+	return lexFile(input, file)
+}
+
+// lex creates a new scanner for the input string.
+// It also takes a file Argument for position handling
+func lexFile(input string, file *token.File) *lexer {
```
[the remark this answers is not in the document]
Comment (slrtbtfs) [reply promql/lex.go]: We need it for the language server.
What was still missing is an exported wrapper for it in the parser.

File: promql/printer.go (hunk 2 above)
Comment (slrtbtfs) [reply promql/printer.go]: Oh

File: promql/lex.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonpasquier) [reply promql/lex.go]: This comment applies to `*lexer.Position()` so it should start like `// Position reports ...`

[the remark this answers is not in the document]
Comment (slrtbtfs) [reply promql/lex.go]: Ah, that makes sense.

**Comment by slrtbtfs** [comment]
The merge conflict should be resolved now.

**Comment by slrtbtfs** [comment]
Is there any specific reason stopping this from being merged?

cc @simonpasquier @krasi-georgiev @brian-brazil @juliusv

File: promql/ast.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brancz) [reply promql/ast.go]: I was just thinking the same thing. I think it's ok to do as a follow up cleanup, but to keep the diff small for this PR, I'd prefer to revert this part.

**Comment by brancz** [comment]
lgtm from an implementation side, @juliusv @brian-brazil @beorn7 are you happy with moving forward with this in general?

As a side note we're going to be investigating a generated parser next, our investigation so far have shown that even if we go with that, the overwhelming majority of this PR would still be required (more comprehensive proposal in progress).

**Comment by juliusv** [comment]
Yeah, I haven't checked the implementation in detail, but in general I'm all for it. I'm very much looking forward to PromQL syntax highlighting and autocompletion :)

File: promql/ast.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (slrtbtfs) [reply promql/ast.go]: I've reverted the change.

**Comment by beorn7** [comment]
I guess @tomwilkie is way more qualified than I am to comment on this.

**Comment by slrtbtfs** [comment]
Here are some benchmarks for this PR.

https://gist.github.com/slrtbtfs/07fec8f8e023e70e50ee220c9bd29ca8

**Comment by slrtbtfs** [comment]
Looks like there is no performance hit. The number of allocations actually seems to have gone slightly down.

**Comment by slrtbtfs** [comment]
The benchmarks above test the whole PromQL engine. Since the parser is only a small part of these workloads,  the changes of this PR might get buried in the noise.

I've added another benchmark (see #6355) , that exclusively tests the parser. Here a negative performance impact is visible, but not dramatic. It can be seen that the position logic adds 5 allocations per parser run.

https://gist.github.com/slrtbtfs/e95ffda4b8cd12b77428d57ac2cdb774

**Comment by brian-brazil** [comment]
I'd consider if the changes are lost in the noise, then all is good.

**Comment by slrtbtfs** [comment]
Closed in favour of adding the features of this PR, after the parser has been replaced by a generated one.

**Pull request closed without merging**